Remove multiple inheritance from menuitemcheckbox and menuitemradio

carmacleod opened the pull request:
**NOTE: June 13, 2020**
For some reason pr-preview did not run, so the Preview and Diff are old (tooltip says "Last updated on May 26, 2020"). Please just look at the code changes in Files changed - there are only 3 changes (2 deletions and an addition).

Closes #1277

- removes checkbox from menuitemcheckbox superclasses (so it only inherits from menuitem)
- removes radio from menuitemradio superclasses (so it only inherits from menuitemcheckbox)
- requires aria-checked on menuitemcheckbox (which menuitemradio inherits)

<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria/pull/1278.html" title="Last updated on May 26, 2020, 12:27 PM UTC (3f845d5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria/1278/3c600ae...3f845d5.html" title="Last updated on May 26, 2020, 12:27 PM UTC (3f845d5)">Diff</a>

jnurthen commented:
I'd prefer to solve this by changing the inheritance of menuitemcheckbox.
To be honest the other states and properties menuitemcheckbox (and eventually menuitemradio) inherit make no sense either:
aria-errormessage, aria-invalid, aria-readonly and aria-required

jnurthen commented:
I'm going to tag the underlying issue for the agenda and if we don't have agreement through github we can discuss on Thursday.

carmacleod commented:
@jnurthen 
> I'd prefer to solve this by changing the inheritance of menuitemcheckbox.
> To be honest the other states and properties menuitemcheckbox (and eventually menuitemradio) inherit make no sense either:
> aria-errormessage, aria-invalid, aria-readonly and aria-required

That's brilliant, and I agree completely, and I will remove `checkbox` from menuitemcheckbox's superclasses in this PR (and add `aria-checked` to the supported attrs). However, I think menuitemcheckbox should still inherit from `menuitem`, so it doesn't solve the `aria-expanded` problem.

(Also, hmmm... maybe `aria-haspopup` should be prohibited on menuitemcheckbox/menuitemradio...).

carmacleod commented:
I guess `radio` should be removed from `menuitemradio`'s superclasses to correspond to removing `checkbox` as a superclass of `menuitemcheckbox`.

carmacleod commented:
This PR:
- Removes checkbox from menuitemcheckbox superclasses (so it only inherits from menuitem)
- Removes radio from menuitemradio superclasses (so it only inherits from menuitemcheckbox)
- Adds aria-checked as required for menuitemcheckbox (which is then inherited by menuitemradio)
- Prohibits aria-expanded and aria-haspopup on menuitemcheckbox and menuitemradio
- Fixes up the javascript so that prohibited attributes (i.e. aria-expanded and aria-haspopup) don't say they are used in or inherit into roles they are prohibited on (i.e. menuitemcheckbox and menuitemradio)
- Adds support for aria-expanded to radio

Since menuitemcheckbox and menuitemradio no longer inherit from an input (i.e. checkbox or radio),
they no longer support aria-errormessage, aria-invalid (2 of the 4 deprecated globals), aria-readonly or aria-required.

carmacleod commented:
From the call:
- don't support aria-expanded on radio
- do support aria-expanded on menuitemradio and menuitemcheckbox
- do remove the multiple inheritance (i.e. remove checkbox from menuitemcheckbox superclasses, and remove radio from menuitemradio superclasses)

carmacleod commented:
Ok, I've made the changes that were discussed on the call, removed the WIP label, and renamed this PR to reflect what it contains now:
- removes the multiple inheritance from menuitemcheckbox and menuitemradio, i.e.
  - removes checkbox from menuitemcheckbox superclasses (so it only inherits from menuitem)
  - removes radio from menuitemradio superclasses (so it only inherits from menuitemcheckbox)
  - requires aria-checked on menuitemcheckbox (which menuitemradio inherits)

This has the desired side effect of removing support for aria-errormessage, aria-invalid, aria-readonly and aria-required from menuitemcheckbox and menuitemradio.

I also moved the javascript to support prohibited non-global attributes to https://github.com/w3c/aria-common/pull/46 to be used later if we need it.

This PR is ready for review.
